third_party/lua/luadebug.lua

```diff
@@ -413,13 +480,148 @@ local function cmd_eval(code)
     -- Call the chunk and collect the results.
     local success, err = pcall(chunk, unpack(rawget(env, "...") or {}))
     if not success then
-        dbg_writeln(color_red("Error:") .. " " .. tostring(err))
+        dbg_write_error(tostring(err))
     end
 
     return false
 end
 
-local function cmd_down()
+--[[
+    Parse source location syntax. One of a kind:
+        luadebug.lua:100
+        luadebug.lua+100
+        :100
+        +100
+--]]
+local function parse_bp(expr)
+    assert(expr)
+    local from, to, file, line = expr:find('^(.-)%s*[+:](%d+)%s*$')
+    if not from then return nil end
+    assert(to == #expr)
+    return file, tonumber(line)
+end
+
+local function _bp_format_expected()
+    dbg_write_error("command expects argument in format filename:NN or " ..
+                    "filename+NN, where NN should be a positive number.")
+end
+
+-- check range to fit in boundaries {from, to}
+local function check_line_max(v, from, to)
+    if type(v) ~= 'number' then
+        dbg_write_error("numeric value expected as line number, " ..
+                        "but received %s", type(v))
+        return false
+    end
+    if v >= from and v <= to then
+        return true
+    end
+    dbg_write_error("line number %d is out of allowed range [%d, %d]",
+                    v, from, to)
+    return false
+end
+
+
+local function cmd_add_breakpoint(bps)
+    assert(bps)
+    local fullfile, line = parse_bp(bps)
+    if not fullfile then
+        _bp_format_expected()
+        return false
+    end
+    if not check_line_max(line, 1, LJ_MAX_LINE) then
+        return false
+    end
+    local info = debug.getinfo(stack_inspect_offset + CMD_STACK_LEVEL, "S")
+    local debuggee = info.source
+    local file = #fullfile > 0 and fullfile or debuggee
+    local normfile = normalize_path(file)
+
+    -- save direct hash (for faster lookup): line -> file
+    if not breakpoints[line] then
+        breakpoints[line] = {}
+    end
+    breakpoints[line][normfile] = true
+    -- save reversed hash information: file -> [lines]
+    if not breakpoints[normfile] then
+        breakpoints[normfile] = {}
+    end
+    breakpoints[normfile][line] = true
```

Comment:
There is a major issue with adding breakpoints by `normfile`, as `normalize_path()` returns just a filename without a path. Thus files `./this/test.lua` and `./other/test.lua` are treated as the same and the added breakpoint will activate on both files.

```
luadebug.lua: Loaded for 2.11.0-entrypoint-695-g562b80988
break via dbg() => ./dbg.lua:8 in chunk at ./dbg.lua:0
   5   
   6   dbg()
   7   
   8=> hello_world()
   9   hello_hell()
luadebug.lua> b test.lua:2
Added breakpoint test.lua:2
   5   
   6   dbg()
   7   
   8=> hello_world()
   9   hello_hell()
luadebug.lua> bl
List of active breakpoints:
	test.lua:2
   5   
   6   dbg()
   7   
   8=> hello_world()
   9   hello_hell()
luadebug.lua> c
/home/ubuntu/Code/review/tarantool-tdbg/build/test.lua:2 in global 'hello_world'
   1   function hello_world()
   2=>●    print('hello world!')
   3   end
luadebug.lua> c
hello world!
/home/ubuntu/Code/review/tarantool-tdbg/build/src/test.lua:2 in global 'hello_hell'
   1   function hello_hell()
   2=>●    print('hello hell')
   3   end
luadebug.lua> c
hello hell
```

Reply:
`gdb` in such cases can create a bp for all files (e.g. `b test.cpp:5`) or a bp for a file by a path (e.g. `b ./one/test.cpp:5`) and handle them correctly:
```
Reading symbols from ./a.out...
(gdb) l
1	void hello_1();
2	void hello_2();
3	
4	int main()
5	{
6		hello_1();
7		hello_2();
8	}
(gdb) b one/test.cpp:4
Breakpoint 1 at 0x11aa: file one/test.cpp, line 5.
(gdb) s
The program is not being run.
(gdb) r
Starting program: /home/ubuntu/tmp/a.out 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, hello_1 () at one/test.cpp:5
5		std::cout << "Hello 1\n";
(gdb) c
Continuing.
Hello 1
hello 2!
[Inferior 1 (process 785145) exited normally]
(gdb) b test.cpp:4
Note: breakpoint 1 also set at pc 0x5555555551aa.
Breakpoint 2 at 0x5555555551aa: test.cpp:4. (2 locations)
(gdb) c
The program is not being run.
(gdb) r
Starting program: /home/ubuntu/tmp/a.out 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, hello_1 () at one/test.cpp:5
5		std::cout << "Hello 1\n";
(gdb) c
Continuing.
Hello 1

Breakpoint 2, hello_2 () at other/test.cpp:5
5		std::cout << "hello 2!\n";
(gdb) c
Continuing.
hello 2!
[Inferior 1 (process 785148) exited normally]
```

Reply:
This can be easily fixed by changing `fio.basename(file)` to `fio.abspath(file)` in `normalize_path()`, this way, a file's breakpoint is saved by full path and cannot collide. I've tested it on the mandelbrot benchmark and it doesn't have any visible slowdown.
```
luadebug.lua: Loaded for 2.11.0-entrypoint-881-gf719e1014
break via dbg() => ./main.lua:8 in chunk at ./main.lua:0
   5   
   6   dbg()
   7   
   8=> hello_1()
   9   hello_2()
luadebug.lua> b ./one/test.lua:2
Added breakpoint /home/ubuntu/Code/review/tarantool-tdbg/build/one/test.lua:2.
   2=>●    print('hello 1!')
   5   
   6   dbg()
   7   
   8=> hello_1()
   9   hello_2()
luadebug.lua> c
/home/ubuntu/Code/review/tarantool-tdbg/build/one/test.lua:2 in global 'hello_1'
   1   function hello_1()
   2=>●    print('hello 1!')
   3   end
luadebug.lua> 
hello 1!
hello 2!
```
The downside to this is that to create a bp user has to provide a relative path to it:
```
luadebug.lua: Loaded for 2.11.0-entrypoint-881-gf719e1014
break via dbg() => ./main.lua:8 in chunk at ./main.lua:0
   5   
   6   dbg()
   7   
   8=> hello_1()
   9   hello_2()
luadebug.lua> b test.lua:2
Added breakpoint /home/ubuntu/Code/review/tarantool-tdbg/build/test.lua:2.
Warning: No source code is available for /home/ubuntu/Code/review/tarantool-tdbg/build/test.lua:2
   5   
   6   dbg()
   7   
   8=> hello_1()
   9   hello_2()
luadebug.lua> c
hello 1!
hello 2!
```

Reply:
* Given your local change please evaluate how it will impact builtin modules debugging? Would `abspath` work in such cases (not, their path is virtual);
* As I already mentioned it in other threads - efficient support of such tail path directories assume to use suffix trees. Ideally such mechanism would provide nice usability gain, but will overcomplicate the code of 1st version.
* Given current status of a patch, and ver small probability to have name confict in real-life scenarios we are targetting today (we do not yet support debugging of Cartridge applications with numerous `init.lua` there, we are rather targetting to debug simple CLI utilities) I'd rather postpone such questionable improvements till next iterations. Until they are asked for.

Reply:
Personally, I would make it in the same way as `gdb` does, having two modes (actually three, considering builtin) - just filename as a `basename()` match and file by path mode, with `abspath()` and builtin check:
1) `luadebug.lua> b file.lua:123` 
2) `luadebug.lua> b ./dir/file.lua:123`
3) `luadebug.lua> b @builtin/file.lua:123`

It doesn't seem too complicated.

Reply:
As we agreed with @igormunkin we do not change (break) current breakpoint behaviour on the regard of using base file name. Eventually, once we get a proper suffix tree, we may improve user experience here.

Reply:
Well, I misunderstood this problem: the one we discussed looks like a feature request **"to set breakpoints by partial paths (e.g. dir/file.lua:123)"**.

This one I also consider as a bug, since something like `src/init.lua` and `src/lib/init.lua` should be considered as different files. Using the same names for different modules is widely used in Lua world, so this should be fixed in scope of this commit. If there is a workaround, please provide one.

Reply:
This is somewhat fixed by suffix tree implementation, see below.